client/cody/src/services/LocalAppDetector.ts

```diff
@@ -2,157 +2,177 @@ import * as vscode from 'vscode'
+        const homeDir = this.localEnv.homeDir
         // if conditions are not met, this will be a noop
-        if (this._watchers.length || !this.isSupported || !this.homeDir || this.isInstalled) {
+        if (!this.isSupported || !homeDir) {
+            debug('LocalAppDetector:init:failed', 'osNotSupported')
             return
         }
         // Create filePaths and file watchers
         const markers = this.localAppMarkers
         for (const marker of markers) {
-            const dirPath = this.expandHomeDir(marker.dir)
+            const dirPath = expandHomeDir(marker.dir, homeDir)
             const dirUri = vscode.Uri.file(dirPath)
             const watchPattern = new vscode.RelativePattern(dirUri, marker.file)
             const watcher = vscode.workspace.createFileSystemWatcher(watchPattern)
-            watcher.onDidChange(() => this.detect())
+            watcher.onDidChange(() => this.fetchApp())
             this._watchers.push(watcher)
             this.appFsPaths.push(dirPath + marker.file)
+            if (marker.hasToken) {
+                this.tokenFsPath = vscode.Uri.file(dirPath + marker.file)
+            }
         }
-        void this.detect()
+        debug('LocalAppDetector:init', 'initialized')
+        await this.fetchApp()
     }
 
-    private async detect(): Promise<void> {
-        if (!this.isSupported || !this.appFsPaths.length) {
+    // Check if App is installed
+    private async fetchApp(): Promise<void> {
+        if (this.localEnv.isAppInstalled || !this.appFsPaths) {
             return
         }
-        if (!this.isInstalled) {
-            if (await Promise.any(this.appFsPaths.map(file => pathExists(file)))) {
-                this.isInstalled = true
-                await this.fetch()
-                this.fire()
-            }
+        debug('LocalAppDetector:fetchApp', 'initializing')
+        if (await Promise.any(this.appFsPaths.map(file => pathExists(file)))) {
+            debug('LocalAppDetector:fetchApp', 'found')
+            this.localEnv.isAppInstalled = true
+            this.appFsPaths = []
+            await this.found('app')
+            await this.fetchToken()
             return
         }
+        debug('LocalAppDetector:detect:fetchApp', 'failed')
     }
 
-    // Check if App is running
-    public async fetch(): Promise<void> {
-        if (!this.isInstalled || this.isRunning) {
+    // Get token from app.json if it exists
+    private async fetchToken(): Promise<void> {
+        if (!this.tokenFsPath || this.localEnv.hasAppJson) {
             return
         }
-        const response = await fetch(`${LOCAL_APP_URL.href}__version`)
-        if (response.status === 200) {
-            this.isRunning = true
-            console.log('App is running.')
+        debug('LocalAppDetector:fetchToken', 'initializing')
+        const data = await vscode.workspace.fs.readFile(this.tokenFsPath).then(data => data.toString())
+        const json = JSON.parse(data) as AppJson
```

Comment:
In case something goes wrong here I suggest that we make this more defensive against errors. Like wrapping it in a function `loadAppJson` which handles when `readFile` fails or when `JSON.parse` fails.
```suggestion
        const appJson = await loadAppJson()
        if (!appJson) {
          return
        }
```
How about:
```ts
async function loadAppJson(uri: Uri): Promise<AppJson | null> {
    try {
        const data = await vscode.workspace.fs.readFile(uri)
        return JSON.parse(data.toString()) as AppJson
    } catch {
        return null
    }
}
```

Reply:
sounds brilliant to me!